# Combining & Exporting Data

## Contains:
### Merging orders and orders_prior 

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#create text file path
path = r'C:\Users\krist\08.2023_InstacartBasketAnalysis'

In [4]:
#import orders_products_prior
df_ords_prior = pd.read_csv(os.path.join(path, 
                                         '02.Data', 
                                         'Original_Data', 
                                         'orders_products_prior.csv'))

In [13]:
#import cleaned orders, used index_col = [0] to remove redundant index column found when visualized orders_clean
df_orders_clean = pd.read_csv(os.path.join(path, 
                                           '02.Data',
                                           'Prepared_Data', 
                                           'orders_checked.csv'), 
                                          index_col = [0])

In [6]:
#visualize part of ords_prior dataset
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [8]:
#check shape of ords_prior
df_ords_prior.shape

(32434489, 4)

In [14]:
#visualize part of orders_clean dataset
df_orders_clean.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  
0                   11.0  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [15]:
#check shape of orders_clean
df_orders_clean.shape

(3421083, 6)

### ords_prior has 32,434,489 rows and 4 columns, orders_clean has 3,421,083 rows and 6 columns 

## Combine orders and orders_prior

In [17]:
#use merge to combine ords_prior with orders_clean using default inner join on order_id
df_merged_large = df_orders_clean.merge(df_ords_prior, 
                                        on = 'order_id', 
                                        indicator = True)

In [18]:
#check output
df_merged_large.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  
0                   11.0         196                  1          0   both  
1                   11.0       14084                  2          0   both  
2                   11.0       12427                  3          0   both  
3                   11.0       26088                  4          0   both  
4                   11.0       26405                  5          0   both

In [19]:
#check to see if full match
df_merged_large['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

### while there appears to be a full match this is because we used inner join. when counts run from outer join there were 32,434,489 both and 206,209 left_only

In [20]:
#check shape of new combined dataset
df_merged_large.shape

(32434489, 10)

### combined dataset has 32,434,489 rows and 10 columns

In [23]:
#export to pkl format
df_merged_large.to_pickle(os.path.join(path, 
                                       '02.Data', 
                                       'Prepared_Data', 
                                       'orders_products_combined.pkl'))